presto-client/src/main/java/io/prestosql/client/auth/external/Tokens.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.client.auth.external;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import net.jodah.failsafe.Failsafe;
+import net.jodah.failsafe.FailsafeException;
+import okhttp3.Call;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.time.Duration;
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static java.net.HttpURLConnection.HTTP_OK;
+import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
+import static java.util.Objects.requireNonNull;
+
+class Tokens
+{
+    public static final int HTTP_FAILED_DEPENDENCY = 424;
+    private final OkHttpClient client;
+    private final ObjectMapper mapper = new ObjectMapper();
+    private Duration timeout;
+
+    Tokens(OkHttpClient client, Duration timeout)
+    {
+        this.client = requireNonNull(client, "client is null");
+        this.timeout = requireNonNull(timeout, "timeout is null");
+    }
+
+    public TokenPoll pollForToken(String tokenUrl)
+            throws TokenPollException
+    {
+        try {
+            return Failsafe.with(RetryPolicies.<TokenPoll>retryUntil(timeout)
+                    .abortIf(TokenPoll::nextTokenUriExists)
+                    .handleResultIf(TokenPoll::isNotAvailableYet)
+                    .handleIf(tx -> false))
```

Comment:
Does it mean we are going to loop even on NPE?

Correct me if I understand, this means that failsafe is not going to assume any exception as failure.